### PR TITLE
[HOPSWORKS-2978] Knative and Kserving metrics

### PR DESCRIPTION
**Pull request by gibchikafa**

### CODE REVIEW

File: attributes/default.rb [hunk 1]
```diff
@@ -90,4 +90,9 @@
 default['cloud']['queue_config']['capacity']             = "10000"
 default['cloud']['queue_config']['max_sample_per_send']  = "5000"
 default['cloud']['queue_config']['batch_send_deadline']  = "60s"
-default['cloud']['metrics']['port']                      = "9096"
\ No newline at end of file
+default['cloud']['metrics']['port']                      = "9096"
+
+default['hopsmonitor']['kube_ca'] = "#{node['certs']['dir']}/kube/kube-ca.cert.pem"
```
Comment (jimdowling) [inline attributes/default.rb:95]: Small thing, but please use spaces to align all of the   = "...." in the file.
See lines 90-93 for example.

File: attributes/default.rb [hunk 2]
```diff
@@ -90,4 +90,9 @@
 default['cloud']['queue_config']['capacity']             = "10000"
 default['cloud']['queue_config']['max_sample_per_send']  = "5000"
 default['cloud']['queue_config']['batch_send_deadline']  = "60s"
-default['cloud']['metrics']['port']                      = "9096"
\ No newline at end of file
+default['cloud']['metrics']['port']                      = "9096"
```
Comment (jimdowling) [inline attributes/default.rb:93]: You are missing at the top of attributes/default.rb (add in line 7)
include_attribute "kube-hops"

If you don't include this, the attributes in lines 95-98 might not find the correct prometheus/root_dir attribute.

File: files/default/dashboards/kserving/knative_control_plane_efficiency.json [hunk 3]
```diff
@@ -0,0 +1,523 @@
+{
```
Comment (jimdowling) [inline files/default/dashboards/kserving/knative_control_plane_efficiency.json:1]: I have no idea, assume this file is correct.

File: recipes/grafana.rb [hunk 4]
```diff
@@ -100,6 +100,18 @@
   mode 0700
 end
 
+directory "#{node['grafana']['base_dir']}/conf/provisioning/dashboards/kserving" do
```
Comment (jimdowling) [inline recipes/grafana.rb:103]: Can you call the directory "kserve" not "kserving"? Both this and the remote directory copied in.

File: recipes/prometheus.rb [hunk 5]
```diff
@@ -91,7 +92,8 @@
       'certificate' => certificate,
       'key' => key,
       'hops_ca' => hops_ca,
-      'managed_cloud' => managed_cloud
+      'managed_cloud' => managed_cloud,
+      'kube_master_ip' => kube_cluster_master_ip,
```
Comment (jimdowling) [inline recipes/prometheus.rb:96]: Remove trailing comma - ','

File: templates/default/dashboards_provisioning.yml.erb [hunk 6]
```diff
@@ -32,3 +32,18 @@ providers:
...
+  type: file
+  # <bool> disable dashboard deletion
+  disableDeletion: false
+  # <bool> enable dashboard editing
+  editable: false
+  # <int> how often Grafana will scan for changed dashboards
+  updateIntervalSeconds: 10
+  options:
+    # <string, required> path to dashboard files on disk. Required
+    path: /srv/hops/grafana/conf/provisioning/dashboards/kserving
```
Comment (jimdowling) [inline templates/default/dashboards_provisioning.yml.erb:48]: kserve

File: templates/default/prometheus.yml.erb [hunk 7]
```diff
@@ -94,7 +94,157 @@ scrape_configs:
...
+        action: keep
+        regex: .+;http-usermetric
+      # Rename metadata labels to be reader friendly
+      - source_labels: [__meta_kubernetes_namespace]
+        target_label: namespace
+      - source_labels: [__meta_kubernetes_pod_name]
+        target_label: pod
+      - source_labels: [__meta_kubernetes_service_name]
+        target_label: service
+<% end -%>
```
Comment (jimdowling) [inline templates/default/prometheus.yml.erb:247]: Chef code looks ok here.

**Review by jimdowling — CHANGES_REQUESTED** [review]
Couple of small changes.

**Comment by jimdowling** [comment]
In Karamelfile, 
add the global dependency to 
recipe: hopsmon::prometheus
    global:
 - kube-hops::hopsmon
 - kabent::default

File: recipes/prometheus.rb [hunk 8]
```diff
@@ -80,19 +80,93 @@
...
+  #Clean up the certs from kubernetes. The JSON.parse does not work so we clean the strings manually
+  kube_data_pattern.each {|k,v| (kube_ca_str).sub!(k,v)}
+  kube_data_pattern.each {|k,v| (kube_key_str).sub!(k,v)}
+  kube_data_pattern.each {|k,v| (kube_crt_str).sub!(k,v)}
+  directory kube_certs_dir do
+    owner node["kagent"]["certs_user"]
+    group node["kagent"]["certs_user"]
+    action :create
+    mode '0700'
+    only_if { node['install']['kubernetes'] == 'true' }
```
Comment (jimdowling) [inline recipes/prometheus.rb:112]: remove

**Review by jimdowling — APPROVED** [review]

File: Berksfile [hunk 9]
```diff
@@ -8,3 +8,4 @@ cookbook 'ndb', github: "logicalclocks/ndb-chef", branch: "master"
 cookbook 'kagent', github: "logicalclocks/kagent-chef", branch: "master"
 cookbook 'tensorflow', github: "logicalclocks/tensorflow-chef", branch: "master"
 cookbook 'java', github: "logicalclocks/java", branch: "v7.0.0-1"
+cookbook 'kube-hops', github: "gibchikafa/kube-hops-chef", branch: "HOPSWORKS-2978"
```
[the remark this answers is not in the document]
Comment (SirOibaf) [reply Berksfile:11]: @gibchikafa - this is still a problem

**Review by SirOibaf — DISMISSED** [review]

File: attributes/default.rb (hunk 2 above)
Comment (jimdowling) [reply attributes/default.rb:93]: Attributes were moved from kube-hops here, no longer relevant.

**Review by jimdowling — APPROVED** [review]

**Pull request merged**